Internally version all detected content by extension

KeyboardNerd opened the pull request:
Conceptual change:
- `Detector` now represents anything that extracts information from a layer blob. `Detector` has two different types for now: `Namespace` ( Detector to extract namespace) and `Feature` ( Detector to extract features )
- `Processor` and `Lister` are renamed to `Detector`

Changes in API:
- Add Detector message type, Namespace message type
- Add detector field to Feature/Namespace to indicate which detector finds a specific feature or namespace
- Change all `scanned_*` to be just `detector`

Changes in Database Model:
- Every features and namespaces are associated with the detectors that found them.
- Aggregated concepts:
  - `Entities`: objects relating the vulnerabilities and the ancestries as the result of Analysis. 
  - `Ancestries`: layers and ancestries scanned with detectors with metadata (name/structure/relation to entities/detectors) saved in the database.
  - `Vulnerabilities`: vulns fetched by the updaters.
  - `Notification`: vulnerability notifications spawned by vulnerability changes.
  - `UpdaterLock`: lock and keyValue.

Changes in Behavior:
- Remove Clair side (dis)enable detectors in configuration. Now, every layer is scanned with every detectors. 

Additional responsibility to Client:
- Client should filter out the features that are not necessary by filtering the detectors ( This responsibility maybe moved to API level on server side in the future )

KeyboardNerd committed:
pgsql: update the query format

KeyboardNerd committed:
config: removed worker config

All processors will now be used to process the layers.

KeyboardNerd committed:
v3: associate feature and namespace with detector

KeyboardNerd commented:
Please review `clair.proto` and `models.go` to confirm the expectation. Those are expected to not be changed in the following commits.

jzelinskie commented on database/detector.go:
```diff
@@ -0,0 +1,53 @@
+package database
```
license header at the top of the file

jzelinskie commented on database/detector.go:
```diff
@@ -0,0 +1,53 @@
+	// Version of the detector
+	Version string
+	// Type of the detector
+	Type DetectorType
```
`type` is a keyword so I'm a little uneasy about using this name. 
As a counter-point, the `reflect` standard library package does define their own `type Type interface {...}`

KeyboardNerd replied:
https://github.com/coreos/clair/blob/master/database/database.go#L42 haha

jzelinskie commented on database/detector.go:
```diff
@@ -0,0 +1,53 @@
+}
+
+// NewNamespaceDetector returns a new namespace detector.
+func NewNamespaceDetector(name string, version string) Detector {
```
`name, verison string`

jzelinskie commented on database/detector.go:
```diff
@@ -0,0 +1,53 @@
+}
+
+// NewFeatureDetector returns a new feature detector.
+func NewFeatureDetector(name string, version string) Detector {
```
`name, version string`

jzelinskie commented on api/v3/clairpb/clair.proto:
```diff
@@ -44,17 +44,39 @@ message Vulnerability {
+  }
+  // The name of the detector.
+  string name = 1;
+  // The Version of the detector.
```
Version -> version

KeyboardNerd committed:
pgsql: Add detector to database schema

'detector' table is added to store the metadata of detectors.
'layer_feature', 'layer_namespace', and 'ancestry_feature' tables are
modified to store the detection relationship between the
feature/namespace with the detector.

KeyboardNerd commented on database/pgsql/migrations/00001_initial_schema.go:
```diff
@@ -14,179 +14,220 @@
+	}
+
+	// detector is analysis extensions used by the worker.
+	detector = dbutil.MigrationQuery{
```
New Table: Detectors

KeyboardNerd replied:
Detector should contain all metadata information of all the extensions used to scan a layer.

KeyboardNerd commented on database/pgsql/migrations/00001_initial_schema.go:
```diff
@@ -14,179 +14,220 @@
+			`CREATE TABLE IF NOT EXISTS layer_detector(
+				id SERIAL PRIMARY KEY,
+				layer_id INT REFERENCES layer ON DELETE CASCADE,
+				detector_id INT REFERENCES detector ON DELETE CASCADE,
```
replaced: Layer Detector now refer to the detector used to scan this layer.

KeyboardNerd commented on database/pgsql/migrations/00001_initial_schema.go:
```diff
@@ -14,179 +14,220 @@
+				id SERIAL PRIMARY KEY,
+				layer_id INT REFERENCES layer ON DELETE CASCADE, 
+				feature_id INT REFERENCES feature ON DELETE CASCADE,
+				detector_id INT REFERENCES detector ON DELETE CASCADE,
```
Layer feature now has reference to detector, which is used to produce this feature. 

The reason for having detector here instead of in `Feature` table is that:
1. Detector only make sense in the context of layer: `Feature`, `Namespace`, and `NamespacedFeature` are also used by the vulnerabilities, which doesn't really care about what scans a feature.
2. Suppose: 
Detector1, Detector2 scans Layer1 and found feature Feature1 by Detector1
Detector1, Detector2 scans Layer2 and found feature Feature1 by Detector2
Since Feature1 is found by Detector1 in Layer1, this relation should be stored somewhere in the database. I think this is the best place.

KeyboardNerd commented on database/pgsql/migrations/00001_initial_schema.go:
```diff
@@ -14,179 +14,220 @@
-	RegisterMigration(migrate.Migration{
-		ID: 1,
-		Up: migrate.Queries([]string{
+var (
```
Schema migration queries are now separated from each other to have a cleaner structure.

KeyboardNerd commented on database/pgsql/migrations/00001_initial_schema.go:
```diff
@@ -14,179 +14,220 @@
+var (
+	// entities are the basic building blocks to relate the vulnerabilities with
+	// the ancestry.
+	entities = dbutil.MigrationQuery{
```
Entities are the *real* immutable content in the database.

On a higher level, Layer ( layer hash with layer content ) is not really immutable because a new extension can scan a layer and add new associated features.

KeyboardNerd commented on ext/featurefmt/driver.go:
```diff
@@ -46,29 +46,33 @@ type Lister interface {
 	RequiredFilenames() []string
+
+	// Version provides the detector's version.
+	Info() database.Detector
```
Please review this design, the detectors should define this Info interface to return the metadata i.e. the name, version, and type of the detector.

KeyboardNerd commented on ext/featurefmt/driver.go:
```diff
@@ -46,29 +46,33 @@ type Lister interface {
-	if name == "" {
-		panic("featurefmt: could not register a Lister with an empty name")
-	}
+func RegisterLister(l Lister) {
```
I will change Lister to detector in the following PRs.

KeyboardNerd commented on database/detector.go:
```diff
@@ -0,0 +1,53 @@
+
+// Detector is an extention to scan a layer's content.
+type Detector struct {
+	// Name of the detector
```
Please review this design to use string to represent the Detector Type.
I can also define `FeatureDetector` and `NamespaceDetector` struct types for the same purpose.

KeyboardNerd commented on config.yaml.sample:
```diff
@@ -51,19 +51,6 @@ clair:
     keyfile:
     certfile:
 
```
User won't be able to define the detectors any more. However, user may want to do the filtering on API level, which means that *maybe* we should change the API to allow the user to query Clair with a set of feature/namespace detectors.

KeyboardNerd commented on cmd/clair/main.go:
```diff
@@ -112,29 +112,16 @@ func configClairVersion(config *Config) {
-	}
-
+	// All listers and detectors are enabled.
+	clair.Processors = database.Processors{Detectors: detectors, Listers: listers}
```
I'll update this in the following PRs.

KeyboardNerd commented on api/v3/clairpb/clair.proto:
```diff
@@ -44,17 +44,39 @@ message Vulnerability {
   repeated Feature affected_versions = 8;
 }
 
+message Detector {
```
Please review: API is changed to have Detector attached to every single feature. Maybe this is not desirable and we should update the RPC endpoint to let the user to specify which detector result they want and whether to return the detectors associated with features.

KeyboardNerd replied:
Keep the API design and don't add additional filtering.

KeyboardNerd reviewed (commented):
Self comments

KeyboardNerd commented on database/models.go:
```diff
@@ -71,6 +65,7 @@ type Layer struct {
 type Namespace struct {
 	Name          string
 	VersionFormat string
+	DetectedBy    Detector
```
Using Detector here maybe not valid. As the reason shows in the database, namespace/feature may be detected by two different detectors in two different layers.

jzelinskie commented on database/pgsql/migrations/00001_initial_schema.go:
```diff
@@ -14,179 +14,220 @@
+	RegisterMigration(dbutil.NewSimpleMigration(1,
+		[]dbutil.MigrationQuery{
+			entities, detector, layer, ancestry,
+			vulnerability, updaterLock, notification,
```
one line each

jzelinskie commented on pkg/dbutil/migrationutil.go:
```diff
@@ -0,0 +1,30 @@
+package dbutil
```
copyright headers

and let's move this into database/psql/migrations

KeyboardNerd committed:
vendor: Add golang-set dependency

Golang-set library is added to make it easier to support set operations.

KeyboardNerd committed:
ext: Lister and Detector returns detector info with detected content

1. Every Lister and Detector are versioned
2. detected content, are returned in a map with detector info as the key

KeyboardNerd committed:
pgsql: Move queries to corresponding files

Aggregate queries in their corresponding files instead of having the
single file for every queries because the database is more complicated.

KeyboardNerd commented on database/models.go:
```diff
@@ -22,47 +22,45 @@ import (
+type Layer struct {
+	Hash             string
+	Namespaces       DetectedNamespaces
+	DetectedFeatures DetectedFeatures
```
DetectedFeatures -> Features

KeyboardNerd committed:
api: Add detectors for RPC

Change the V3 implementation to accommondate the detectors.

KeyboardNerd committed:
clair: Implement worker detector support

The worker is changed to accommodate the new database model and API.
Worker is refactored to move the database query helper functions to pkg.

KeyboardNerd commented on database/models.go:
```diff
@@ -19,50 +19,121 @@ import (
-// ProcessedBy.
+// DetectedNamespaces contains a map from detector to the detected namespace.
+// Nil detected namespace means it's not found.
+type DetectedNamespaces map[Detector]*Namespace
```
The new map type introduces a lot more complexity. It makes the worker complicated and also the database implementation. Maybe, we don't want this map to organize since the map will at most contain around 1000 features.

KeyboardNerd replied:
Flatten the map to be []DetectedNamespace, which contains Detector and Namespace.
This is for simplifying the data structure, we can then use pointers and a new data structure if we need to improve the performance.

KeyboardNerd commented on api/v3/clairpb/clair.proto:
```diff
@@ -44,17 +44,42 @@ message Vulnerability {
 
+message Detector {
+  enum Type {
+    DETECTOR_TYPE_INVALID = 0;
```
This name is a requirement by the prototool linter.

KeyboardNerd commented on api/v3/util.go:
```diff
@@ -34,41 +33,50 @@ func GetClairStatus(store database.Datastore) (*pb.ClairStatus, error) {
 // GetPbAncestryLayer retrieves an ancestry layer with vulnerabilities and
 // features in an ancestry based on the provided database layer.
-func GetPbAncestryLayer(session database.Session, layer database.AncestryLayer) (*pb.GetAncestryResponse_AncestryLayer, error) {
+func GetPbAncestryLayer(tx database.Session, layer database.AncestryLayer) (*pb.GetAncestryResponse_AncestryLayer, error) {
```
This change is for consistency of naming database.Session variable as tx ( transaction )

KeyboardNerd commented on api/v3/util.go:
```diff
@@ -34,41 +33,50 @@ func GetClairStatus(store database.Datastore) (*pb.ClairStatus, error) {
+	// this feature for now, we should refactor the implementation if there's
+	// any performance issue. It's expected that the number of features is less
+	// than 1000.
+
```
nit: remove line

KeyboardNerd commented on api/v3/util.go:
```diff
@@ -34,41 +33,50 @@ func GetClairStatus(store database.Datastore) (*pb.ClairStatus, error) {
+
+	for _, feature := range affectedFeatures {
 		if !feature.Valid {
 			return nil, status.Error(codes.Internal, "ancestry feature is not found")
```
not related to this change: If a feature is not valid, this should indicate a bigger problem.

KeyboardNerd commented on cmd/clair/main.go:
```diff
@@ -102,40 +102,13 @@ func stopCPUProfiling(f *os.File) {
-	clair.EnabledUpdaters = strutil.CompareStringListsInBoth(config.Updater.EnabledUpdaters, updaters)
+		"Detectors": strings.Join(database.SerializeDetectors(clair.EnabledDetectors), ","),
+		"Updaters":  strings.Join(clair.EnabledUpdaters, ","),
+	}).Info("Enabled Clair components")
```
nit: lower case

KeyboardNerd commented on config.yaml.sample:
```diff
@@ -51,19 +51,6 @@ clair:
-      - apk
-      - dpkg
-      - rpm
-
```
now the lister/detectors are not configurable

KeyboardNerd commented on database/database.go:
```diff
@@ -99,6 +107,9 @@ type Session interface {
 	// namespaced features. If the ancestry is not found, return false.
 	FindAncestry(name string) (ancestry Ancestry, found bool, err error)
 
+	// PersistDetector inserts a slice of detectors if not in the database.
```
todo: define the error handling behavior of all the interfaces

KeyboardNerd commented on database/database.go:
```diff
@@ -120,12 +131,10 @@ type Session interface {
 
-	// PersistLayer persists a layer's content in the database. The given
-	// namespaces and features can be partial content of this layer.
+	// PersistLayer append a layer's content in the database.
```
appends

KeyboardNerd commented on database/database.go:
```diff
@@ -120,12 +131,10 @@ type Session interface {
-	// The layer, namespaces and features are expected to be already existing
-	// in the database.
-	PersistLayer(hash string, namespaces []Namespace, features []Feature, processedBy Processors) error
+	// If any feature or namespace or detector are not in the database, it returns not found error.
```
If any feature, namespace, or detector is ...

KeyboardNerd commented on database/database.go:
```diff
@@ -157,8 +166,8 @@ type Session interface {
-	// and their pages are specified by the paination token, which, if empty, are
-	// always considered first page.
+	// and their pages are specified by the pagination token, which, if empty,
+	// are always considered first page.
```
...by the pagination token, which should be considered first page when it's empty.

KeyboardNerd commented on database/detector.go:
```diff
@@ -0,0 +1,144 @@
+
+// DetectorTypes contains all detector types.
+var DetectorTypes = []DetectorType{
+	NamespaceType,
```
This is being used as database.NamespaceType, which doesn't make sense. 
Rename to DetectorNamespaceType

KeyboardNerd commented on database/detector.go:
```diff
@@ -0,0 +1,144 @@
+
+// ErrFailedToParseDetectorType is the error returned when a detector type could
+// not be parsed from a string.
+var ErrFailedToParseDetectorType = errors.New("failed to parse DetectorType from input")
```
Merge vars 
var ( ...
)

KeyboardNerd commented on database/models.go:
```diff
@@ -22,47 +22,128 @@ import (
-
-	// DetectedFeatures are the features introduced by this layer when it was
+	// Hash is the name of the ancestry's layer.
+	Hash string
```
TODO: We need to address that Hash MUST be a content addressable hash for the layer tar ball.

KeyboardNerd replied:
TODO: Add specification: hash must be non-empty

jzelinskie replied:
This is comment also wrong. Hash is the sha256 digest of the manifest for a particular image.

KeyboardNerd replied:
This comment is wrong, Hash is on the Ancestry Layer NOT the Ancestry.

KeyboardNerd commented on database/models.go:
```diff
@@ -22,47 +22,128 @@ import (
+// Layer is a layer with all the detected features and namespaces.
 type Layer struct {
-	LayerMetadata
+	Hash string
```
^ same here

KeyboardNerd commented on database/models.go:
```diff
@@ -22,47 +22,128 @@ import (
+	return namespaces
+}
+
+// TODO(sidac): change the name to DetectedNamespace
```
Really?

KeyboardNerd replied:
No, don't change the name

KeyboardNerd commented on database/models.go:
```diff
@@ -85,6 +166,11 @@ type Feature struct {
 }
 
+// NewFeature creates a feature.
+func NewFeature(Name, Version, VersionFormat string) Feature {
```
Is this really used?

KeyboardNerd commented on database/models.go:
```diff
@@ -201,7 +287,7 @@ type VulnerabilityNotificationWithVulnerable struct {
 }
 
-// MetadataMap is for storing the metadata returned by vulnerability database.
+// MetadataMap is for storing the metadata returned by vulnerability
```
revert this line

KeyboardNerd commented on database/pgsql/ancestry_test.go:
```diff
@@ -15,198 +15,127 @@
 func TestFindAncestry(t *testing.T) {
 	store, tx := openSessionForTest(t, "FindAncestry", true)
 	defer closeTest(t, store, tx)
+	// Reference: https://github.com/golang/go/wiki/TableDrivenTests
```
remove this?

KeyboardNerd commented on database/pgsql/feature.go:
```diff
@@ -27,10 +27,53 @@ import (
 )
 
+const (
+	// feature.go
```
remove

KeyboardNerd commented on database/pgsql/feature.go:
```diff
@@ -27,10 +27,53 @@ import (
 	errFeatureNotFound = errors.New("Feature not found")
 )
 
+type feature struct {
```
is this being used?

KeyboardNerd commented on database/pgsql/detector.go:
```diff
@@ -0,0 +1,197 @@
+	return detectors, nil
+}
+
+func (tx *pgSession) findAllDetectors() (detectorMap, error) {
```
TODO: we should have a cache layer for the detector

KeyboardNerd commented on database/pgsql/layer.go:
```diff
@@ -18,300 +18,349 @@ import (
+func sanitizePersistLayerInput(hash string, features []database.LayerFeature, namespaces []database.LayerNamespace, detectedBy []database.Detector) error {
 	if hash == "" {
-		return -1, commonerr.NewBadRequestError("Empty Layer Hash is not allowed")
+		return commonerr.NewBadRequestError("expected non-empty layer hash")
```
TODO: in the future ( not this PR ), we should update the way we handle the errors.

KeyboardNerd commented on database/pgsql/layer.go:
```diff
@@ -18,300 +18,349 @@ import (
-		keys[i*2+1] = d
+	// TODO(sidac): This kind of type conversion is very useless and wasteful,
+	// we need interfaces around the database models to reduce these kind of
+	// operations.
```
TODO: Yeah, rework the interface is very important, the current interface is bad. For example, the strong assumption that the features should be first inserted, and the layers MUST relate to the same features found by the same metadata! Bad.

KeyboardNerd commented on database/pgsql/layer.go:
```diff
@@ -18,300 +18,349 @@ import (
-	for i, nsID := range ids {
-		keys[i*2] = id
-		keys[i*2+1] = nsID
+	keys := make([]interface{}, len(features)*3)
```
TODO: abstract this, it's hard for me to use!

KeyboardNerd commented on database/pgsql/migrations/00001_initial_schema.go:
```diff
@@ -14,179 +14,223 @@
+				id SERIAL PRIMARY KEY,
+				ancestry_id INT REFERENCES ancestry ON DELETE CASCADE,
+				ancestry_index INT NOT NULL,
+				layer_id INT NOT NULL REFERENCES layer ON DELETE RESTRICT,
```
NOT NULL is not necessary

KeyboardNerd commented on database/pgsql/namespace.go:
```diff
@@ -23,10 +23,19 @@ import (
 	errNamespaceNotFound = errors.New("Requested Namespace is not in database")
 )
 
+type namespace struct {
```
is this used?

KeyboardNerd commented on database/pgsql/namespace.go:
```diff
@@ -23,10 +23,19 @@ import (
+)
+
 var (
 	errNamespaceNotFound = errors.New("Requested Namespace is not in database")
```
Why do we have this unexported thing here??

KeyboardNerd commented on database/pgsql/notification.go:
```diff
@@ -26,6 +26,50 @@ import (
 )
 
+const (
+	// notification.go
```
remove

KeyboardNerd commented on database/pgsql/vulnerability.go:
```diff
@@ -26,6 +26,73 @@ import (
 )
 
+const (
+	lockVulnerabilityAffects = `LOCK vulnerability_affected_namespaced_feature IN SHARE ROW EXCLUSIVE MODE`
```
TODO: is this query even necessary??

KeyboardNerd commented on ext/featurens/driver.go:
```diff
@@ -61,60 +67,69 @@ func RegisterDetector(name string, d Detector) {
+	return namespaces, nil
 }
 
 // RequiredFilenames returns the total list of files required for all
```
update the comment

KeyboardNerd commented on ext/featurens/driver_test.go:
```diff
@@ -3,11 +3,12 @@ package featurens_test
 import (
 	"testing"
 
+	"github.com/coreos/clair/ext/featurens"
```
re order

KeyboardNerd commented on pkg/dbutil/dbutil.go:
```diff
@@ -0,0 +1,299 @@
+	return tx.FindAncestry(name)
+}
+
+func FindLayer(datastore database.Datastore, hash string) (layer database.Layer, ok bool, err error) {
```
comment

KeyboardNerd commented on pkg/dbutil/dbutil.go:
```diff
@@ -0,0 +1,299 @@
+	return nil
+}
+
+func CacheRelatedVulnerability(datastore database.Datastore, features []database.NamespacedFeature) error {
```
TODO: I start to think that the abstraction on database layer is wrong. Somehow we need to figure out a way to get ride of the implicit database feature relationship.

KeyboardNerd commented on pkg/dbutil/dbutil.go:
```diff
@@ -0,0 +1,299 @@
+	return FilterDetectors(detectors, database.NamespaceType)
+}
+
+func FilterDetectors(detectors []database.Detector, dtype database.DetectorType) []database.Detector {
```
is this being used? remove

KeyboardNerd commented on database/pgsql/migrations/00001_initial_schema.go:
```diff
@@ -22,170 +22,170 @@ func init() {
-		name TEXT NULL,
-		version_format TEXT,
-		UNIQUE (name, version_format));`,
+				id SERIAL PRIMARY KEY,
```
Since this file is replaced with later commit, we can remove this commit.

KeyboardNerd replied:
don't remove

KeyboardNerd commented on api/v3/clairpb/clair.proto:
```diff
@@ -100,12 +123,11 @@ message GetAncestryResponse {
-    repeated string scanned_listers = 4;
-    // The configured list of namespace detectors used to scan an ancestry.
-    repeated string scanned_detectors = 5;
+    // The detectors used to scan this Ancestry. It maybe not the current set of
```
Rephrase: The detectors used to scan this Ancestry. It may not be the same set of detectors in the Clair instance.

KeyboardNerd commented on api/v3/util.go:
```diff
@@ -50,7 +50,11 @@ func GetPbAncestryLayer(session database.Session, layer database.AncestryLayer)
+	// this feature for now, we should refactor the implementation if there's
+	// any performance issue. It's expected that the number of features is less
+	// than 1000.
+	for _, feature := range affectedFeatures {
```
This is caused by a wrongly resolved merge conflict, move this to proper commit.

KeyboardNerd replied:
it's fine

KeyboardNerd commented on ext/featurefmt/driver.go:
```diff
@@ -48,13 +47,19 @@ type Lister interface {
 	RequiredFilenames() []string
 }
 
+type lister struct {
```
TODO: It makes me think if we should keep the name "lister" but to name database.Detector to database.ExtensionInfo or database.ExtensionMetadata.

Lister: finds features
Detector: finds namespaces
ExtensionMetadata: description of lister or detector ( or updater? )

KeyboardNerd replied:
don't do it the name is changed, it's changed.

KeyboardNerd commented on ext/featurens/driver.go:
```diff
@@ -29,7 +29,7 @@ import (
 
 var (
 	detectorsM sync.RWMutex
```
Not related: the mutex is right now useless, should we remove it and only when we decide to implement runtime reloading of the lister/detector, we should add it again.

KeyboardNerd commented on ext/imagefmt/driver.go:
```diff
@@ -127,21 +128,23 @@ func Extract(format, path string, headers map[string]string, toExtract []string)
 
 		layerReader = r.Body
 	} else {
+		log.WithField("path", strutil.CleanURL(path)).Debug("start reading layer blob from local file system...")
```
This is the preliminary step to improve the overall logging infra, which is basically non-existing right now.

KeyboardNerd commented on worker.go:
```diff
@@ -60,296 +48,179 @@ type LayerRequest struct {
-	features    []database.Feature
-
-	err error
+// processResult stores layer's content detected by `processedBy` processors.
```
update the comment

KeyboardNerd commented on worker.go:
```diff
@@ -60,296 +48,179 @@ type LayerRequest struct {
-		}
+		log.WithField("layer", req.Hash).Debug("found no existing layer in database")
 
+		layer := database.Layer{Hash: req.Hash}
```
we can inline this to
`existingLayer: database.Layer{Hash: req.Hash}`

KeyboardNerd commented on worker.go:
```diff
@@ -60,296 +48,179 @@ type LayerRequest struct {
-	}
-	for ns := range mapNS {
-		namespaces = append(namespaces, ns)
+func getProcessResultLayers(results map[string]*processResult) map[string]database.Layer {
```
this function merges the layer existing in database and the newly scanned result.

KeyboardNerd commented on worker.go:
```diff
@@ -60,296 +48,179 @@ type LayerRequest struct {
-	}
-	if !ok {
-		return false, nil
+	ancestry, ok, err := dbutil.FindAncestry(datastore, name)
```
It's wasteful to find ancestry and all the features associated here. We should later consider to have this function on database interface.

KeyboardNerd commented on worker.go:
```diff
@@ -360,166 +231,114 @@ func ProcessAncestry(datastore database.Datastore, imageFormat, name string, lay
 	features := map[string]map[string]introducedFeature{}
 	ancestryLayers := []database.AncestryLayer{}
 	for index, layer := range layers {
 		// Initialize the ancestry Layer
```
this comment is unnecessary

KeyboardNerd reviewed (commented):
Note to other PR reviewer
TODO means it's a thing for the future PR not this one
NIT means it's not blocking merging the PR
Not related means it's a old issue and can be not addressed in this PR
Otherwise it means blocking the PR.

jzelinskie commented on api/v3/clairpb/convert.go:
```diff
@@ -122,23 +122,38 @@ func VulnerabilityWithFixedInFromDatabaseModel(dbVuln database.VulnerabilityWith
+	return &Detector{
+		Name:    detector.Name,
+		Version: detector.Version,
+		Type:    Detector_Type(Detector_Type_value[string(detector.DType)]),
```
Wow this kinda sucks. I wonder if gogo protobuf generates something more ergonomic than this.

jzelinskie commented on database/detector.go:
```diff
@@ -0,0 +1,144 @@
+	// NamespaceDetectorType is a type of detector that extracts the namespaces.
+	NamespaceDetectorType DetectorType = "DETECTOR_TYPE_NAMESPACE"
+	// FeatureDetectorType is a type of detector that extracts the features.
+	FeatureDetectorType DetectorType = "DETECTOR_TYPE_FEATURE"
```
Is there a reason to even use a string? Why not have an integer?

```go
const (
  NamespaceDectectorType DetectorType = iota
  FeatureDetectorType
  ...
)
```

KeyboardNerd replied:
because I want to store enum in the database. Well I can write a conversion to make this work anyway.

jzelinskie replied:
Enums are most often just integers in the database that have special meaning for the application. You use enums because it makes the column more simple (thus easier to migrate, copy, compare values etc...)

KeyboardNerd replied:
Yes, that's the point isn't it. DetectorType (Detector Enum) is just like Severity Enum. You can look up the database schema for that.

jzelinskie commented on database/detector.go:
```diff
@@ -0,0 +1,144 @@
+}
+
+// SerializeDetectors returns the string representation of given detectors.
+func SerializeDetectors(detectors []Detector) []string {
```
We could put this in strutil if you change the function to be:

```go
func StringifyList(xs []fmt.Stringer) (ys []string) {
  for _, x := range xs {
    ys = append(ys, x.String())
  }
  return
}
```

KeyboardNerd replied:
ok, then it will require the type conversion from []database.Detector to []fmt.Stringer, which sucks

jzelinskie replied:
Yeah. This is super unfortunate... It's because interfaces are different memory layouts and creating a list of them is O(N), so you have to explicitly make the copy yourself rather than having Go do it for you when you pass it into the function.

KeyboardNerd replied:
It makes me think if this kind of design is wrong after all. I think we are thinking about functions using *generics* but interfaces are not generics.

KeyboardNerd replied:
I think I should keep the original design for the serialization.

jzelinskie replied:
:thumbsup:

jzelinskie commented on database/pgsql/detector.go:
```diff
@@ -0,0 +1,197 @@
+
+		if err := tx.persistLayerDetector(layerID, id); err != nil {
+			return err
+		}
```
Looks like you're forgetting to insert into the set.

jzelinskie commented on pkg/dbutil/dbutil.go:
```diff
@@ -0,0 +1,288 @@
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package dbutil
```
I think everything in here makes sense to just have in the database package rather than in a utility package.

jzelinskie commented on pkg/testutil/testutil.go:
```diff
@@ -0,0 +1,285 @@
+package testutil
```
I think all of this is useful for only the database. I think this would also be fine being just in the database package.

Also, copyright header

KeyboardNerd replied:
Yes, those are for comparing the database models *but* it's not only for the database tests. It's also used in the worker test and in the future updater test.

jzelinskie reviewed (changes requested):
This is my first pass. I'm going to need to take a look much harder at the database changes.

KeyboardNerd committed:
database: move dbutil and testutil to database from pkg

Move dbutil and testutil to database from pkg
Rename all "result"

jzelinskie commented on cmd/clair/main.go:
```diff
@@ -180,6 +154,18 @@ func Boot(config *Config) {
 	st.Stop()
 }
 
+func initLogger(flagLogLevel *string) {
```
I avoid using the word `init` to avoid confusion around the keyword.

jzelinskie commented on cmd/clair/main.go:
```diff
@@ -102,40 +102,13 @@ func stopCPUProfiling(f *os.File) {
-
-	clair.EnabledUpdaters = strutil.CompareStringListsInBoth(config.Updater.EnabledUpdaters, updaters)
+		"Detectors": strings.Join(database.SerializeDetectors(clair.EnabledDetectors), ","),
+		"Updaters":  strings.Join(clair.EnabledUpdaters, ","),
```
Do these have to be strings? I'd be nice to keep them as arrays in the JSON logs.
Also the detectors string is borderline unreadable:

```json
{"Detectors":"DETECTOR_TYPE_FEATUREDetector/apk/1.0,DETECTOR_TYPE_FEATUREDetector/dpkg/1.0,DETECTOR_TYPE_FEATUREDetector/rpm/1.0,DETECTOR_TYPE_NAMESPACEDetector/alpine-release/1.0,DETECTOR_TYPE_NAMESPACEDetector/apt-sources/1.0,DETECTOR_TYPE_NAMESPACEDetector/lsb-release/1.0,DETECTOR_TYPE_NAMESPACEDetector/os-release/1.0,DETECTOR_TYPE_NAMESPACEDetector/redhat-release/1.0"}
```

jzelinskie commented on database/dbutil.go:
```diff
@@ -0,0 +1,286 @@
+		nsSet.Add(ns)
+	}
+
+	result := make([]Namespace, 0, nsSet.Cardinality())
```
result is always a bad variable name

jzelinskie commented on database/dbutil.go:
```diff
@@ -0,0 +1,286 @@
+		fSet.Add(f)
+	}
+
+	result := make([]Feature, 0, fSet.Cardinality())
```
result is always a bad variable name

jzelinskie commented on database/dbutil.go:
```diff
@@ -0,0 +1,286 @@
+
+// PersistNamespaces wraps session PersistNamespaces function with begin and
+// commit.
+func PersistNamespaces(datastore Datastore, namespaces []Namespace) error {
```
We should come up with a naming convention for the functions that are just wrapping calls in a transaction.

jzelinskie commented on database/dbutil.go:
```diff
@@ -0,0 +1,286 @@
+	}
+
+	inter := d1Set.Intersect(d2Set)
+	result := make([]Detector, 0, inter.Cardinality())
```
result is always a bad variable name

jzelinskie commented on database/dbutil.go:
```diff
@@ -0,0 +1,286 @@
+		bySet.Add(d)
+	}
+
+	for _, feature := range new.Features {
```
new is a keyword

jzelinskie commented on database/detector.go:
```diff
@@ -0,0 +1,144 @@
+		}
+	}
+
+	return "", ErrFailedToParseSeverity
```
copypasta?

jzelinskie commented on database/detector.go:
```diff
@@ -0,0 +1,144 @@
+
+// String returns a unique string representation of the detector.
+func (d Detector) String() string {
+	return fmt.Sprintf("%sDetector/%s/%s", d.DType, d.Name, d.Version)
```
I comment on main, but yeah this is pretty verbose and hard to read

jzelinskie commented on database/models.go:
```diff
@@ -22,47 +22,129 @@ import (
-
 // Ancestry is a manifest that keeps all layers in an image in order.
 type Ancestry struct {
+	// Name is expected to be the sha-256 digest of the manifest for a
```
Name is a globally unique value for a set of layers. This is often the sha256 digest of an OCI/Docker manifest.

jzelinskie reviewed (changes requested):
I think the general approach for the database is good. We discussed it enough beforehand that it's pretty much exactly what I expected. I think that there were a bunch of things I noticed for database optimizations, but I don't think we should really spend time on that until we're positive it's the finalized schema.

some additional thoughts... i ran the codebase and sat pondering our current schema for a bit and came up with a few things that might not necessarily need to be done in this PR:

### Get rid of namespaced_feature

ancestry_feature -> (id, ancestry_layer_id, feat_id, feat_detector_id, ns_id, ns_detector_id)

### make kv table consistent

- rename fields to `updater/$NAMESPACE`
- prefix value to include the SCM: `git-sha:hf92dk9`

### detector type

- rename column type -> dtype
- change column to this enum: `CREATE TYPE detector_type AS ENUM ('feature', 'namespace');`

### debian:unstable ns needs to be mapped to a number

### we should lstrip "v" off of the alpine namespace version numbers, no other namespaces do that

KeyboardNerd committed:
database: Update database model and interface for detectors

All detected features and namespaces under the context of Layer and
Ancestry will now have the detectors associated, so that the API can
provide the detection information to the Client.

KeyboardNerd committed:
pgsql: Implement database queries for detector relationship

* Refactor layer and ancestry
* Add tests
* Fix bugs introduced when the queries were moved

KeyboardNerd commented:
This is done: 
detector type
rename column type -> dtype
change column to this enum: CREATE TYPE detector_type AS ENUM ('feature', 'namespace');

KeyboardNerd committed:
database: rename utility functions with commit/rollback

All database utility functions are renamed to explicitly say if it will
commit changes or rollback changes on success.

KeyboardNerd committed:
api: Rename detector type to DType

Rename detector type to DType because all reserved key words should be
avoided used as type name or variable name.
